ZelBack/src/services/utils/pathSecurity.js

```diff
@@ -0,0 +1,354 @@
+ * This is used in non-strict mode for basic traversal detection.
+ *
+ * @param {string} component - Path component to check
+ * @returns {boolean} True if component is a traversal attempt
+ */
+function isTraversalComponent(component) {
+  return component === '.' || component === '..';
+}
+
+/**
+ * Basic path validation (non-strict mode).
+ * Only blocks obvious dangerous patterns without character restrictions.
+ *
+ * @param {string} userPath - User-provided path
+ * @throws {Error} If path contains traversal attempts
+ */
+function validatePathBasic(userPath) {
+  if (!userPath || typeof userPath !== 'string') {
+    return;
+  }
+
+  const normalizedPath = normalizePathSeparators(userPath);
+  const components = normalizedPath.split('/');
+
+  for (const component of components) {
+    if (isTraversalComponent(component)) {
+      throw new Error('Invalid path: directory traversal not allowed');
+    }
+  }
+}
+
+/**
+ * Sanitize and validate a relative path to prevent directory traversal.
+ * Uses defense-in-depth with both blocklist and allowlist validation.
+ * Ensures the resolved path stays within the allowed base directory.
+ *
+ * @param {string} userPath - User-provided relative path
+ * @param {string} basePath - The allowed base directory
+ * @param {object} options - Optional configuration
+ * @param {boolean} options.strict - If true, enforce strict allowlist (default: true)
+ * @returns {string} Resolved safe absolute path
+ * @throws {Error} If path attempts to escape base directory or contains invalid characters
+ *
+ * @example
+ * // Valid usage
+ * sanitizePath('appdata', '/apps/myapp')  // Returns '/apps/myapp/appdata'
+ * sanitizePath('logs/app.log', '/apps/myapp')  // Returns '/apps/myapp/logs/app.log'
+ * sanitizePath('', '/apps/myapp')  // Returns '/apps/myapp'
+ *
+ * // Throws error
+ * sanitizePath('..', '/apps/myapp')  // Error: directory traversal
+ * sanitizePath('../other', '/apps/myapp')  // Error: directory traversal
+ * sanitizePath('foo\0bar', '/apps/myapp')  // Error: null byte
+ */
+function sanitizePath(userPath, basePath, options = {}) {
+  const { strict = true } = options;
+
+  // If no user path provided, return the base path
+  if (!userPath || typeof userPath !== 'string') {
+    return basePath;
+  }
+
+  // === LAYER 1: Blocklist checks (fast rejection of known-bad patterns) ===
+
+  // Check for null bytes (can bypass security checks in some systems)
+  if (userPath.includes('\0')) {
+    throw new Error('Invalid path: null bytes not allowed');
+  }
+
+  // Normalize backslashes to forward slashes BEFORE any path operations
+  // This ensures consistent handling and prevents bypass via backslashes
+  const normalizedUserPath = normalizePathSeparators(userPath);
+
+  // Check for traversal patterns in the normalized path
+  // We check each component separately to catch '..' as a path component
+  const components = normalizedUserPath.split('/');
```

Comment:
The only check here that really matters is this one:

resolvedPath !== normalizedBase && !resolvedPath.startsWith(normalizedBase + path.sep

The rest is just code bloat

Assuming that we have already rejected if it contains backslashes. We could also use `path.normalize` instead of reinventing the wheel.

Can also do something as simple as this:

const relative = path.relative(parent, dir);
const isRelative = relative && !relative.startsWith('..') && !path.isAbsolute(relative);